heads.go

```diff
@@ -99,63 +104,104 @@ func (hh *heads) Replace(h, c cid.Cid, height uint64) error {
 		return err
 	}
 
+	hh.cacheMux.Lock()
+	defer hh.cacheMux.Unlock()
+
+	if !batching {
+		hh.cache[c] = height
+	}
+
 	err = hh.delete(store, h)
 	if err != nil {
 		return err
 	}
+	if !batching {
+		delete(hh.cache, h)
+	}
 
 	if batching {
 		err := store.(ds.Batch).Commit()
 		if err != nil {
 			return err
 		}
+		delete(hh.cache, h)
+		hh.cache[c] = height
 	}
 	return nil
 }
 
 func (hh *heads) Add(c cid.Cid, height uint64) error {
 	hh.logger.Infof("adding new DAG head: %s (height: %d)", c, height)
-	return hh.write(hh.store, c, height)
+	if err := hh.write(hh.store, c, height); err != nil {
+		return err
+	}
+
+	hh.cacheMux.Lock()
+	{
+		hh.cache[c] = height
+	}
+	hh.cacheMux.Unlock()
+	return nil
 }
 
 // List returns the list of current heads plus the max height.
 func (hh *heads) List() ([]cid.Cid, uint64, error) {
+	var maxHeight uint64
+	var heads []cid.Cid
+
+	hh.cacheMux.RLock()
+	{
+		heads = make([]cid.Cid, 0, len(hh.cache))
+		for head, height := range hh.cache {
+			heads = append(heads, head)
+			if height > maxHeight {
+				maxHeight = height
+			}
+		}
+	}
+	hh.cacheMux.RUnlock()
+
+	sort.Slice(heads, func(i, j int) bool {
+		ci := heads[i].Bytes()
+		cj := heads[j].Bytes()
+		return bytes.Compare(ci, cj) < 0
+	})
+
+	return heads, maxHeight, nil
+}
+
+func (hh *heads) primeCache() (ret error) {
 	q := query.Query{
 		Prefix:   hh.namespace.String(),
 		KeysOnly: false,
 	}
 
 	results, err := hh.store.Query(q)
 	if err != nil {
-		return nil, 0, err
+		return err
 	}
 	defer results.Close()
 
-	heads := make([]cid.Cid, 0)
-	var maxHeight uint64
 	for r := range results.Next() {
 		if r.Error != nil {
-			return nil, 0, r.Error
+			return r.Error
 		}
 		headKey := ds.NewKey(strings.TrimPrefix(r.Key, hh.namespace.String()))
 		headCid, err := dshelp.DsKeyToCidV1(headKey, cid.DagProtobuf)
 		if err != nil {
-			return nil, 0, err
+			return err
 		}
 		height, n := binary.Uvarint(r.Value)
 		if n <= 0 {
-			return nil, 0, errors.New("error decocding height")
+			return errors.New("error decoding height")
 		}
-		heads = append(heads, headCid)
-		if height > maxHeight {
-			maxHeight = height
+
+		hh.cacheMux.Lock()
```

Comment:
Given we are initializing, we don't even need to lock, do we? Nothing would be racy against the execution of this. And if we lock we can just do it once and defer unlock?